### ISSUE TITLE
Node Discovery Optimizations for Granting

### ISSUE THREAD

**Issue by KPrasch** [nucypher/nucypher#2911]
### Porter Network Request Lifecycle

A recap of Porter's network requests chronologically:

- GET `/sample`
- Porter Samples
    - RPC Call to `getActiveStakingProviders` get list of `n` eth addresses
    - `block_until_number_of_nodes_is` == `n` else (10s timeout):
        - Fetch Certificate
        - `/public_information`
    - If sampled is known (10s timeout):
        - Fetch Certificate
        - GET `/ping`
    - If not known:
        - Get another value from Pool
- Porter responds to GET Request to `/sample`

##### Cheaper

- [x] Extend timeouts
- [ ] Min/Max Porter samples
- [x] Optimistic TLS certificate caching (#2908)
- [ ] Optimizing the worker pool
- [ ] Use a formal concurrency library
- [ ] Option to skip liveliness check (Porter)?
- [ ] Cache node liveliness

##### More Expensive

**Comment by fjarri** [comment]
My five cents. Some of it I outlined in #2876

What I do in `nucypher-async`:

# Learning

The learner maintains two databases: one of verified nodes, and one of urls where a valid node *may* be found. Namely, the second database contains correspondences of staking provider address to pairs `(host, port)` (referred to as a "contact"), and (optionally, if known) the stake.

Differences from the reference:
- General error tolerance: if there's something wrong during verification (of an existing node or a contact), the corresponding entry is discarded.
- When learning from another node, the teacher's metadata returned as a part of the pack is compared to what we have on file; if they're different, the teacher node is discarded from the verified database. (NOTE: in that case the received metadata is still incorporated in the database; maybe it shouldn't be).
- For the rest of the nodes in the pack, the metadata fields besides `host`, `port` and `stakind_provider_address` are discarded, and they're stored in the contacts database.
- During "passive learning" (when another node sends my node some metadata), all of it is discarded except for the entries with the same host as the sender's remote address. Those metadatas are further stripped to `host`, `port` and `stakind_provider_address` as above and added to the contact database. That leaves the possibility of another node spamming contacts with the same host but different ports; not sure how to avoid that.
- Each verified node in the database gets a re-verification time; a queue of scheduled verifications is maintained, and when an event comes up, the node is re-verified.
- Currently, the re-verification is scheduled for 1 hour at first, with the timeout increasing x1.5 every time it's verified, or set to the time when the node's certificate expires, whatever comes earlier. Other criteria may be added.
- If there are entries in the contacts database, they're scheduled for verification right away (in batches, with a 5s delay between batches). Currently random ones are taken, but it may be better to prioritize the ones reported by more nodes.
- Learning is a separate task scheduled for every 90 seconds using a random verified node. Again, some non-trivial strategy is possible (e.g. taking the one that wasn't learned from for the longest period).
- Verification and learning are concurrent and reentrable, each one locks a contact (host/port) it uses, so other tasks cannot touch it. It seems that 10-20 concurrent verifications are possible to eliminate I/O downitme.
- The stake size of the node is queried as a part of the verification. Also a `getActiveStakingProviders()` call is run once a day, providing info for the stakes of unverified contacts (necessary for Alice's sampling).


# Querying Ursulas

## Cold start

If Alice or Bob want to get some nodes without using Porter, they need to get some initial data in their Learner instances. I am not sure how to best approach this; for now I require a "seed round" to be executed which verifies a seed node, gets a list of contacts, and calls `getActiveStakingProviders()`. This gives us an idea of where to look for if we want specific nodes, or nodes we sampled based on stake sizes. Bob really does not need the `getActiveStakingProviders()`, he's only interested in contacts, since he knows the staking provider addresses he needs to query.


## Querying known addresses

This is always done by Bob, or by Alice if `handpicked_ursulas` were provided. In Alices case, for simplicity, I first try to verify all `handpicked_ursulas`, and then sample the remaining nodes. This can be merged to be done in parallel.


## Sampling based on stake

Decision: how recent should the information about the stake be? I am querying `getActiveStakingProviders()` every day, plus the stake for specific nodes is obtained during verification. Are there any timeouts on unstaking/changing stake? This can be incorporated into scheduling the next re-verification time.

Decision: what qualifies a live node? For now I'm defining it as "has been verified within a minute". I don't think pinging is necessary if that's the case.

Decision: if a sampled node is not confirmed live (e.g. was verified too long ago), do we wait for verification or try another one? The way I treat it now is to have an `overhead>=0`. I sample the reservoir, and if the node is live, I `yield` it; if not, a concurrent task is started for re-verification. I stop drawing when I reach `quantity + overhead`. Then I wait on the channel where verification tasks drop their results. If a result comes in, and it's verified, I `yield` it, if it's a failure, I draw one more node, check for liveness and possibly start a verification task.

(Note here that each verification task reports to the verified node database; that verification status is not lost).


## Scheduling re-verification in Porter

If there are many users issuing requests all the time the re-verifications above will be enough to keep the response time low. But if the requests are sparse, we may want to schedule verification depending on the stake size, that is do it more often than an Ursula would.
